### PR TITLE
Re-implement Tracking CSE and Greedy Rewriter using listeners

### PR DESCRIPTION
**Pull request by Mogball**
Depends on https://github.com/google/iree-llvm-sandbox/pull/139

The tracking CSE and greedy rewriter are reimplemented with a listener.

### CODE REVIEW

File: lib/Dialects/LinalgTransform/Transforms/TrackingCSE.cpp [hunk 1]
```diff
@@ -7,296 +7,16 @@
...
-void TrackingCSE::simplify(Operation *root) {
-  for (Region &region : root->getRegions()) {
-    ScopedMapTy knownValues;
-    simplifyRegion(knownValues, region);
-  }
-
-  for (Operation *op : opsToErase)
-    op->erase();
-}
+using namespace mlir::linalg;
```
Comment (ftynse) [inline lib/Dialects/LinalgTransform/Transforms/TrackingCSE.cpp:15]: Why is this needed?

File: include/Dialects/LinalgTransform/TrackingCSE.h [hunk 2]
```diff
@@ -9,13 +9,11 @@
 #ifndef MLIR_DIALECT_LINALG_TRANSFORMS_TRACKINGCSE_H
 #define MLIR_DIALECT_LINALG_TRANSFORMS_TRACKINGCSE_H
 
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
```
Comment (ftynse) [inline include/Dialects/LinalgTransform/TrackingCSE.h:12]: I wouldn't want this dependency unless strictly necessary.

**Review by ftynse — COMMENTED** [review]
LGTM

File: include/Dialects/LinalgTransform/TrackingCSE.h (hunk 2 above)
Comment (Mogball) [reply include/Dialects/LinalgTransform/TrackingCSE.h:12]: `SmallVector<LinalgOp>` requires the type to be complete, so unfortunately a forward declaration won't work.
